### ISSUE TITLE
LDAP - Synchronize issue

### ISSUE THREAD

**Issue by ijafif** [snipe/snipe-it#1230]
Greetings from Seychelles !!!

I have set up SNIPE-IT on our Intranet server but are unable to import users via LDAP even if the LDAP file was configured.

<?php

return array(

    /*
    |--------------------------------------------------------------------------
    | URL
    |--------------------------------------------------------------------------
    |
    | URL for the LDAP server. This should start with ldap://, for example:
    | ldap://ldap.yourserver.com
    |
    */
    'url' => "LDAP://SCAA-DC3.scaa.sc:389",

    /*
    |--------------------------------------------------------------------------
    | Username
    |--------------------------------------------------------------------------
    |
    | Username to use to connect authenticate to LDAP, for example:
    | cn=read-only-admin,dc=example,dc=com
    |
    */
	'username' => "CN=Administrator,CN=Users,DC=scaa,DC=sc",

	/*
    |--------------------------------------------------------------------------
    | Password
    |--------------------------------------------------------------------------
    |
    | Password to use when authenticating to LDAP.
    |
    */
	'password' => "********",

	/*
    |--------------------------------------------------------------------------
    | Basedn
    |--------------------------------------------------------------------------
    |
    | The base where the search for users will be executed, for example:
    | dc=example,dc=com
    |
    */
	'basedn'   => "dc=scaa,dc=sc",

	/*
    |--------------------------------------------------------------------------
    | Filter
    |--------------------------------------------------------------------------
    |
    | The search filter for the LDAP query. This probably does not have to be
    | changed.
    |
    */
	'filter' => "&(cn=*)",

	/*
    |--------------------------------------------------------------------------
    | LDAP field names that will be retrieved to create a user.
    |
    | Using the username as an example:
    | If I set 'result.username' => 'my-org-username', the code will connect to
    | LDAP as follows (where $results[$i] represents a row in the LDAP query:
    | $username-to-insert-in-snipe-it = $results[$i]["my-org-username"][0]
    |
    |--------------------------------------------------------------------------
    |
    | The search filter for the LDAP query.
    | Note: all these fields are required, and they should all be lowercase.
    |
    */
    'result.username' => "",
    'result.last.name'  => "",
    'result.first.name' => "",

    /*
    | These fields are optional as not all LDAP directories will have it.  If yours
    | does not have them, just leave these blank and the extra check will
    | be omitted.
    */
    'result.active.flag' => "",
    'result.emp.num'  => "",
    'result.email' => "",

    /*
    |--------------------------------------------------------------------------
    | LDAP filter query for authentication
    |--------------------------------------------------------------------------
    |
    | The LDAP query that we want to execute when authenticating a user. This
    | should not have to be changed.
    |
    */
    'authentication.filter.query' => "uid=",

    /*
    |--------------------------------------------------------------------------
    | LDAP Version
    |--------------------------------------------------------------------------
    |
    | Version of LDAP you are using.
    |
    */
    'version' => 3,

);

Can anybody assist?

I am contactable via email (iafif@scaa.sc) 

Thank You and anticipating a reply soon.

Kind Regards,

Idris Jamil Afif
System Analyst & Website Administrator
Seychelles Civil Aviation Authority

**Comment by ddreier** [comment]
You need to populate the LDAP field names. Here's what mine look like (for Active Directory):

    'result.username' => "samaccountname",
    'result.emp.num'  => "employeeid",
    'result.last.name'  => "sn",
    'result.first.name' => "givenname",
    'result.email' => "mail",

FWIW, Active Directory doesn't seem to populate the `employeeid` field, so you can probably just leave that one blank.

**Comment by ijafif** [comment]
Dear Daniel,

Thank You for responding - Here's my new configuration:

    |--------------------------------------------------------------------------
    |
    | The search filter for the LDAP query.
    | Note: all these fields are required, and they should all be lowercase.
    |
    */
    'result.username' => "administrator@scaa.sc",
    'result.last.name'  => "administrator",
    'result.first.name' => "scaa",

    /*
    | These fields are optional as not all LDAP directories will have it.  If yours
    | does not have them, just leave these blank and the extra check will
    | be omitted.
    */
    'result.active.flag' => "",
    'result.emp.num'  => "",
    'result.email' => "",

Which results in only one location being in the drop-down list (we have more locations in our Active Directory)

I get this on Snipe-IT LDAP page:

Validator failed: 
•The firstname field is required.
•The lastname field is required.
•The username field is required.

**Comment by ddreier** [comment]
If you're using Active directory, those fields should read exactly the same as in my example.

"samaccountname" is the username property in Active Directory, "sn" is the last name property, "givenname" is the first name property, and "mail" is the email property.

**Comment by ijafif** [comment]
Dear Daniel,

Seems to work but now I am getting the Validator failed message:

Validator failed: 
•The firstname field is required.
•The lastname field is required.
•The username field is required.

**Comment by ddreier** [comment]
Hmm, can you provide a screenshot of where you're getting that error? Possibly another copy of your LDAP config as well.

**Comment by ijafif** [comment]
Dear Daniel,

Here you go for the LDAP config:

<?php

return array(

    /*
    |--------------------------------------------------------------------------
    | URL
    |--------------------------------------------------------------------------
    |
    | URL for the LDAP server. This should start with ldap://, for example:
    | ldap://ldap.yourserver.com
    |
    */
    'url' => "LDAP://SCAA-DC3.scaa.sc:389",


    /*
    |--------------------------------------------------------------------------
    | Username
    |--------------------------------------------------------------------------
    |
    | Username to use to connect authenticate to LDAP, for example:
    | cn=read-only-admin,dc=example,dc=com
    |
    */
	'username' => "CN=Administrator,CN=Users,DC=scaa,DC=sc",

	/*
    |--------------------------------------------------------------------------
    | Password
    |--------------------------------------------------------------------------
    |
    | Password to use when authenticating to LDAP.
    |
    */
	'password' => "SCAAits0",

	/*
    |--------------------------------------------------------------------------
    | Basedn
    |--------------------------------------------------------------------------
    |
    | The base where the search for users will be executed, for example:
    | dc=example,dc=com
    |
    */
	'basedn'   => "dc=scaa,dc=sc",


	/*
    |--------------------------------------------------------------------------
    | Filter
    |--------------------------------------------------------------------------
    |
    | The search filter for the LDAP query. This probably does not have to be
    | changed.
    |
    */
	'filter' => "&(cn=*)",


	/*
    |--------------------------------------------------------------------------
    | LDAP field names that will be retrieved to create a user.
    |
    | Using the username as an example:
    | If I set 'result.username' => 'my-org-username', the code will connect to
    | LDAP as follows (where $results[$i] represents a row in the LDAP query:
    | $username-to-insert-in-snipe-it = $results[$i]["my-org-username"][0]
    |
    |--------------------------------------------------------------------------
    |
    | The search filter for the LDAP query.
    | Note: all these fields are required, and they should all be lowercase.
    |
    */
    'result.username' => "administrator",
    'result.last.name'  => "",
    'result.first.name' => "",

    /*
    | These fields are optional as not all LDAP directories will have it.  If yours
    | does not have them, just leave these blank and the extra check will
    | be omitted.
    */
    'result.active.flag' => "",
    'result.emp.num'  => "",
    'result.email' => "Administrator@scaa.sc",

    /*
    |--------------------------------------------------------------------------
    | LDAP filter query for authentication
    |--------------------------------------------------------------------------
    |
    | The LDAP query that we want to execute when authenticating a user. This
    | should not have to be changed.
    |
    */
    'authentication.filter.query' => "uid=",

    /*
    |--------------------------------------------------------------------------
    | LDAP Version
    |--------------------------------------------------------------------------
    |
    | Version of LDAP you are using.
    |
    */
    'version' => 3,

);

**Comment by ijafif** [comment]
Screenshot: 
![untitled](https://cloud.githubusercontent.com/assets/15008863/10385698/b1a2fa26-6e5d-11e5-8ff2-20420ec59826.jpg)

**Comment by ddreier** [comment]
In your ldap.php, result.username, result.last.name, and result.first.name are still incorrect.

If you're using Active Directory, result.username should be set to "samaccountname", result.last.name should be "sn", and result.first.name should be "givenname".

**Comment by ijafif** [comment]
Dear Daniel,

I changed again:

    | The search filter for the LDAP query.
    | Note: all these fields are required, and they should all be lowercase.
    |
    */
    'result.username' => "Administrator",
    'result.last.name'  => "<not set>",
    'result.first.name' => "<not set>",

Still no luck

**Comment by ddreier** [comment]
So...What is your LDAP directory?

**Comment by ijafif** [comment]
I don't get you there

**Comment by ddreier** [comment]
What is running on SCAA-DC3.scaa.sc? Is it an Active Directory Domain Controller?

**Comment by ijafif** [comment]
Yes, it is our Active Directory Domain Controller

**Comment by ddreier** [comment]
Then this section of ldap.php needs to look *exactly* like this:
````
| The search filter for the LDAP query.
| Note: all these fields are required, and they should all be lowercase.
|
*/
'result.username' => "samaccountname",
'result.last.name'  => "sn",
'result.first.name' => "givenname",
````

**Comment by ijafif** [comment]
Let me try again and send you the code to verify

**Comment by ijafif** [comment]
<?php

return array(

    /*
    |--------------------------------------------------------------------------
    | URL
    |--------------------------------------------------------------------------
    |
    | URL for the LDAP server. This should start with ldap://, for example:
    | ldap://ldap.yourserver.com
    |
    */
    'url' => "LDAP://SCAA-DC3.scaa.sc:389",


    /*
    |--------------------------------------------------------------------------
    | Username
    |--------------------------------------------------------------------------
    |
    | Username to use to connect authenticate to LDAP, for example:
    | cn=read-only-admin,dc=example,dc=com
    |
    */
	'username' => "CN=Administrator,CN=Users,DC=scaa,DC=sc",

	/*
    |--------------------------------------------------------------------------
    | Password
    |--------------------------------------------------------------------------
    |
    | Password to use when authenticating to LDAP.
    |
    */
	'password' => "SCAAits0",

	/*
    |--------------------------------------------------------------------------
    | Basedn
    |--------------------------------------------------------------------------
    |
    | The base where the search for users will be executed, for example:
    | dc=example,dc=com
    |
    */
	'basedn'   => "dc=scaa,dc=sc",


	/*
    |--------------------------------------------------------------------------
    | Filter
    |--------------------------------------------------------------------------
    |
    | The search filter for the LDAP query. This probably does not have to be
    | changed.
    |
    */
	'filter' => "&(cn=*)",


	/*
    |--------------------------------------------------------------------------
    | LDAP field names that will be retrieved to create a user.
    |
    | Using the username as an example:
    | If I set 'result.username' => 'my-org-username', the code will connect to
    | LDAP as follows (where $results[$i] represents a row in the LDAP query:
    | $username-to-insert-in-snipe-it = $results[$i]["my-org-username"][0]
    |
    |--------------------------------------------------------------------------
    |
    | The search filter for the LDAP query.
    | Note: all these fields are required, and they should all be lowercase.
    |
    */
    'result.username' => "samaccountname",
    'result.last.name'  => "sn",
    'result.first.name' => "givenname",

    /*
    | These fields are optional as not all LDAP directories will have it.  If yours
    | does not have them, just leave these blank and the extra check will
    | be omitted.
    */
    'result.active.flag' => "",
    'result.emp.num'  => "",
    'result.email' => "Administrator@scaa.sc",

    /*
    |--------------------------------------------------------------------------
    | LDAP filter query for authentication
    |--------------------------------------------------------------------------
    |
    | The LDAP query that we want to execute when authenticating a user. This
    | should not have to be changed.
    |
    */
    'authentication.filter.query' => "uid=",

    /*
    |--------------------------------------------------------------------------
    | LDAP Version
    |--------------------------------------------------------------------------
    |
    | Version of LDAP you are using.
    |
    */
    'version' => 3,


);

**Comment by ddreier** [comment]
You just posted your Domain Administrator password. I'd delete that very quickly.

**Comment by ddreier** [comment]
Otherwise, change this and it should be good:
````
'result.email' => "mail",
````

**Comment by ijafif** [comment]
It's bound to change and so will do - Please delete your last comment too :-)

**Comment by ijafif** [comment]
Dear Daniel,

At long last, it works :+1:

I restarted our associated web services and logged back into Snipe-IT

Thank You for your patience and help :+1: 

Have A Blessed Weekend

**Comment by ddreier** [comment]
Glad I could help. Please close the GitHub issue since it's resolved.

**Comment by mahesh2471990** [comment]
@ddreier 
Hello Daniel

![ldap](https://cloud.githubusercontent.com/assets/16305213/11891827/4ed5a3aa-a588-11e5-9529-309effc69f17.JPG)

**Comment by mahesh2471990** [comment]
But having Some Error.

![error](https://cloud.githubusercontent.com/assets/16305213/11891833/61a2efce-a588-11e5-9f71-9dd503dfc431.JPG)

**Comment by ddreier** [comment]
@mahesh2471990 Please open a separate issue.
